atomic_reactor/plugins/pre_resolve_module_compose.py

```diff
@@ -0,0 +1,272 @@
+        self.stream = stream
+        self.version = version
+        self.mmd = mmd
+
+
+class ComposeInfo(object):
+    def __init__(self, compose_id, base_module, modules, repo_url):
+        self.compose_id = compose_id
+        self.base_module = base_module
+        self.modules = modules
+        self.repo_url = repo_url
+
+
+WORKSPACE_SOURCE_KEY = 'compose_info'
+
+
+def get_compose_info(workflow):
+    key = ResolveModuleComposePlugin.key
+    if key not in workflow.plugin_workspace:
+        return None
+    return workflow.plugin_workspace[key].get(WORKSPACE_SOURCE_KEY, None)
+
+
+def set_compose_info(workflow, source):
+    key = ResolveModuleComposePlugin.key
+
+    workflow.plugin_workspace.setdefault(key, {})
+    workspace = workflow.plugin_workspace[key]
+    workspace[WORKSPACE_SOURCE_KEY] = source
+
+
+class ODCSClient(object):
+    def __init__(self, url, insecure=False, token=None, log=None):
+        if url.endswith('/'):
+            self.url = url
+        else:
+            self.url = url + '/'
+        self.insecure = insecure
+        self.token = token
+        self.log = log
+
+    def _auth_headers(self):
+        headers = {}
+        if self.token:
+            headers['OIDC_access_token'] = self.token
+            headers['OIDC_CLAIM_scope'] = ('openid https://id.fedoraproject.org/scope/groups '
+                                           'https://pagure.io/odcs/new-compose '
+                                           'https://pagure.io/odcs/renew-compose '
+                                           'https://pagure.io/odcs/delete-compose')
+            # FIXME: passing the claim scope in the request doesn't make any sense,
+            #   because we're not to be trusted for what scopes we've claimed, but
+            #   match what the server wants.
+
+        return headers
+
+    def start_compose(self, source_type, source):
+        body = {
+            'source': {
+                'type': source_type,
+                'source': source
+            }
+        }
+
+        self.log.info("Starting compose for source_type={source_type}, source={source}"
+                      .format(source_type=source_type, source=source))
+        response = requests.post(self.url + 'composes/',
+                                 json=body,
+                                 headers=self._auth_headers())
+        response.raise_for_status()
+
+        return response.json()
+
+    def wait_for_compose(self, compose_id,
+                         burst_retry=1,
+                         burst_length=30,
+                         slow_retry=10,
+                         timeout=300):
+        self.log.debug("Getting compose information for information for compose_id={}"
+                       .format(compose_id))
+        url = self.url + 'composes/' + str(compose_id)
+        headers = self._auth_headers()
+        start_time = time.time()
+        while True:
+            response = requests.get(url, headers=headers)
```

Comment:
Instead of using `requests` directly, `from atomic_reactor.util import get_retrying_requests_session` and use e.g.
```
session = get_retrying_requests_session()
response = session.get(...)
```

That way, retries will be handled transparently (for retryable requests).

Reply:
Hmm, this is not retrying on failure, it is polling against the json content of successful requests. I don't think there's any way obvious way to use the urllib3 Retry functionality for this. [I also would be reluctant to have exponential back-off here... I think it would cause unnecessary slow waits.] The individual requests could be retried, of course.

Reply:
We do also need to retry on failure though, to allow for ODCS to be offline for short periods.
http://osbs.readthedocs.io/en/latest/contributors.html#retry-on-error